src/schematics/ng-add/add-scripts.ts

```diff
@@ -0,0 +1,111 @@
+ * user will be able to run them in parallel. Created scripts will be:
+ *
+ * - build:single-spa:first-cool-app
+ * - serve:single-spa:first-cool-app
+ *
+ * - build:single-spa:second-cool-app
+ * - serve:single-spa:second-cool-app
+ */
+export function addScripts(
+  tree: Tree,
+  pkgPath: string,
+  pkg: any,
+  project: string | undefined,
+): void {
+  if (project) {
+    addScriptsForTheSpecificProject(pkg, project);
+  } else {
+    addDefaultScripts(pkg);
+  }
+
+  tree.overwrite(pkgPath, JSON.stringify(pkg, null, 2));
+}
+
+function addScriptsForTheSpecificProject(pkg: any, project: string): void {
+  const port = parseExistingScriptsAndChoosePort(pkg.scripts);
+
+  pkg.scripts[
+    `build:single-spa:${project}`
+  ] = `ng build ${project} --prod --deploy-url http://localhost:${port}/`;
+
+  pkg.scripts[
+    `serve:single-spa:${project}`
+  ] = `ng s --project ${project} --disable-host-check --port ${port} --deploy-url http://localhost:${port}/ --live-reload false`;
+}
+
+/**
+ * In that case the user didn't provide any `--project` argument, that probably means
+ * that he has a single project in his workspace and we want to provide a default script.
+ */
+function addDefaultScripts(pkg: any): void {
+  pkg.scripts[
+    'build:single-spa'
+  ] = `ng build --prod --deploy-url http://localhost:${DEFAULT_PORT}/`;
+
+  pkg.scripts[
+    'serve:single-spa'
+  ] = `ng s --disable-host-check --port ${DEFAULT_PORT} --deploy-url http://localhost:${DEFAULT_PORT}/ --live-reload false`;
+}
+
+function parseExistingScriptsAndChoosePort(scripts: any): number {
+  const collectedScripts: string[] = collectExistingServeSingleSpaScripts(scripts);
+  // For instance `[4200, 4201, 4202]`.
+  const ports: number[] = getPortsFromCollectedScripts(collectedScripts);
+
+  if (ports.length === 0) {
+    return DEFAULT_PORT;
+  }
+
+  const lastPort = ports.pop();
+  // `4202 + 1 -> 4203` next port for the new project.
+  return lastPort! + 1;
+}
+
+function collectExistingServeSingleSpaScripts(scripts: any): string[] {
+  const collectedScripts: string[] = [];
+
+  for (const key of Object.keys(scripts)) {
+    if (key.startsWith('serve:single-spa')) {
+      collectedScripts.push(scripts[key]);
+    }
+  }
+
+  return collectedScripts;
+}
+
+function getPortsFromCollectedScripts(collectedScripts: string[]): number[] {
+  const ports: number[] = [];
+
+  for (const script of collectedScripts) {
+    const match: RegExpMatchArray | null = script.match(/--port \d{4}/);
+
+    if (match === null) {
+      continue;
+    }
+
+    // `match[0]` will be a string e.g. `--port 4200`,
+    // we split by space to get that `4200`.
+    const [, port] = match[0].split(' ');
+    ports.push(+port);
+  }
+
+  // Sorts all numbers for ascending order. For example we will get
+  // `[4200, 4201, 4202]` sorted numbers. We will need `4202` to get next port.
+  return ports.sort((a, b) => a - b);
```

Comment:
The sort function here is unnecessary, right? Each port is already a number due to line 105, so default sort will work?

```suggestion
  return ports.sort();
```

Reply:
Honestly saying haven't used `sort` in practice, just googled for `sort ascending order` :smile:

Reply:
This is where I got it https://stackoverflow.com/questions/1063007/how-to-sort-an-array-of-integers-correctly

Reply:
@joeldenning upd.

Reply:
Looks like it still has the unnecessary sort function?

Reply:
Does it not work if you call sort() without the callback? If it works let’s remove it

Reply:
removed